### ISSUE TITLE
`docker ps` unresponsive until mounting of volume via volume driver finishes

### ISSUE THREAD

**Issue by itamarst** [moby/moby#18504]
Docker v1.9.1 on CentOS.

Scenario is as follows; `docker run --volume-driver flocker -v myvolume:/data busybox` running in background and in parallel a `docker ps` call. Both take about 20 seconds to finish. During this time the following happens in this order:
1. `VolumeDriver.Create` is sent by Docker, succeeds with appropriate response.
2. `VolumeDriver.Path` is sent by Docker, presumably on behalf of `docker ps`. Flocker docker plugin returns an error since mount path is not yet available. I am guessing `Create` and `Path` can't run in parallel because of locks in the volume store but am not familiar enough with the code paths to be sure.
3. `VolumeDriver.Mount` is sent by Docker, no response is received for a while (creating and attaching EBS volumes can take upwards of 20 seconds).
4. The ongoing `docker ps` never prints anything, which presumably means it hasn't received a response.... until about the same time `VolumeDriver.Mount` finishes, at which point it lists the new container.

I have verified that indeed `GET /v1.21/containers/json` does not return a result until the container is up and running. The strong implication (without reading through the code) is that looking up containers blocks on `VolumeDriver.Mount` finishing. This is a problem insofar as the latter can take a while.

**Comment by itamarst** [comment]
```
[root@ip-172-31-1-217 ~]# docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:25:01 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:25:01 UTC 2015
 OS/Arch:      linux/amd64
[root@ip-172-31-1-217 ~]# docker info
Containers: 20
Images: 38
Server Version: 1.9.1
Storage Driver: devicemapper
 Pool Name: docker-202:1-18147011-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: 
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 740.2 MB
 Data Space Total: 107.4 GB
 Data Space Available: 6.49 GB
 Metadata Space Used: 2.007 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.10.0-229.14.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
CPUs: 2
Total Memory: 6.897 GiB
Name: ip-172-31-1-217
ID: BRJB:WB5M:KK53:2J5R:5W5V:HXRD:XOAC:EMTH:5HZ7:J3AQ:WZO5:KXXE
Debug mode (server): true
 File Descriptors: 30
 Goroutines: 41
 System Time: 2015-12-08T19:39:27.918020864Z
 EventsListeners: 0
 Init SHA1: ac8f1bfea8577235f1e9c2dfaace54207d64ee49
 Init Path: /usr/libexec/docker/dockerinit
 Docker Root Dir: /var/lib/docker
[root@ip-172-31-1-217 ~]# uname -a
Linux ip-172-31-1-217 3.10.0-229.14.1.el7.x86_64 #1 SMP Tue Sep 15 15:05:51 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by thaJeztah** [comment]
Thanks for the report @itamarst 

ping @calavera @cpuguy83 for volumes/volume plugins

**Comment by itamarst** [comment]
I would describe this as a bug report, not an enhancement request: Docker becoming unresponsive has lots of unfortunate side-effects for callers.

**Comment by thaJeztah** [comment]
Hm, yeah, not sure why I did that.

**Comment by LK4D4** [comment]
@cpuguy83 Do you know if this was fixed?

**Comment by cpuguy83** [comment]
Yes, probably still a problem as the container lock is held while we are mounting volumes.

**Comment by LLParse** [comment]
I can confirm this is still a problem in version `1.12.3`.

**Comment by cpuguy83** [comment]
This is not a problem since ~17.05 since `docker ps` is populated from a snapshot instead of the live object.
There are still problems with other endpoints like `docker inspect`, but `docker ps` is not a problem.
